### DISCUSSION TITLE
Attack speed

### DISCUSSION THREAD

**Discussion by Deax7** [trickerer/Trinity-Bots#932] [General]
Playing on Ashen Repack with these bots and they have correct attack speed in description (bot-info) but they ignore variable weapon speed (typical for vanilla) and their offhand attacks seems to have some type of cd. Is that case for bots in general? It makes fast weapons completely unviable and even haste less desirable.

There is maybe a icd to attacks so mainhand attacks and offhand attacks cannot occur in ~same time window (for technical reasons). Attacks seems to be following one another and are not independent.

Extra:
Settings for enchants, poisons, pets.. does not save and are set back to default when login.

.npcbot order cast... often fails to do anything without any error message.

Pets often fails to attack. Pets have very low damage (often single digits)

Some abilities ignore role of the player. (Paladin casting Salvation on player tank)

We are unable to set specific auras to be autocasted.

As i said. Iam playing on Ashen Order Repack and these things can be specific for their custom needs. Thx for insight.

**Comment by trickerer** [comment]
Off-hand attack delay was probably introduced in https://github.com/azerothcore/azerothcore-wotlk/commit/6e4a9bbb


https://github.com/trickerer/AzerothCore-wotlk-with-NPCBots/commit/5c4085e95a1c20053cd7b2840b0af5b4a8c1a362
https://github.com/trickerer/AzerothCore-wotlk-with-NPCBots/commit/e15689cf521d22b1ffea5f5abd23020f0d8ccd65
https://github.com/trickerer/AzerothCore-wotlk-with-NPCBots/commit/3fbdf3b781e921ed0a8c3df39eb98a5c2362058f

**Reply by Deax7** [reply]
Oh, thx for reply!
So on your end everything is fine?
For example rogue with fast 1.5 speed weapons in both hands and some haste attacking exactly like it should be? 
Seems like no form of haste works on them (from item procs, talents)
Like i can get bot to 0.5s attack speed for each hand, testing it with skada doesnt reflect this kind of attack speed

**Reply by trickerer** [reply]
> So on your end everything is fine?

On my end I don't experience that because: a) I don't really play the game, b) I mainly work on TrinityCore. That's why it is new for me after 2 month since that commit
We'll see if I can make an exception for bots so their off-hand attacks are delayed in a way similar to player's

**Reply by Deax7** [reply]
ok, i will let you cook :) the problem is also with attack speed of weapons (2h, 1h, dual w, ranged) or feral cats in general ignoring haste effects somehow and fast weapons being subpar in general for bots coz all of these issues

thats all from me, thank you for your good work :)

**Comment by trickerer** [comment]
So, I did some AC testing.

The commit I mentioned only adds delay to the first off-hand attack and this code only gets executed once at attack start (confirmed using debugger).

Then I tested the number of melee swings bot gets while dual-wielding. Here warrior bot uses 2.6 speed weapon in main hand and 1.90 in off-hand. No other equipment, no flurry, no blood frenzy. 2 minute segments

- Raw damage output (no abilities, no haste)

![d1_3](https://github.com/user-attachments/assets/f219fb42-610a-460c-85af-58c983fc2655)
![d1_1](https://github.com/user-attachments/assets/44600f62-4515-43de-8260-8f66dd617f69)
![d1_2](https://github.com/user-attachments/assets/3d6b3d09-4583-48e2-9ec7-260b1db0b22d)

108 melee swings, no additional delay on off-hand attack is visible, 2 off-hand attacks before next main hand swing are happening periodically as expected
Is it the correct though? We expect about `120 / (((2.6 + 1.9) / 2) / 2) = 106.667` attacks so 108 is OK (this segment is 121 seconds).

- No abilities, 100% haste

![d2_3](https://github.com/user-attachments/assets/89801938-2576-4049-9027-e466489cd5ad)
![d2_1](https://github.com/user-attachments/assets/189fb396-1adf-409a-8384-92a977c0eb65)
![d2_2](https://github.com/user-attachments/assets/63289e97-8e75-4e57-92ce-421da2fefb1b)

211 swings, about the exact number we expect (106.667 * 2 ~= 213), so haste effect was not suppressed or ignored in any way

- 100% haste, no attack timer resetting abilities (slam, heroic throw)

![d3_1](https://github.com/user-attachments/assets/b9326da4-a39d-4e50-9409-bdb5ae217744)
![d3_2](https://github.com/user-attachments/assets/2ccab20b-0e65-40fc-b4e8-f43b0ef5fb31)
![d3_3](https://github.com/user-attachments/assets/6e27c8d1-5e12-4e4c-baeb-ca11a16c6968)

211 swings again so using abilities did not interfere with haste at all

Bots also properly benefit from haste rating from both spells and gear. Here bot receives 500 haste rating from a spell (13494) and 313 haste rating from gear

![d4](https://github.com/user-attachments/assets/c64590d1-cc31-46bc-8483-5aa30c6cbf25)

Conclusion: I'll have to assume any problems with haste you're having come from your repack.
